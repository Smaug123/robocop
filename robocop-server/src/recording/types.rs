use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct RecordedEvent {
    pub timestamp: String,      // ISO 8601 timestamp
    pub correlation_id: String, // Unique ID to group related events
    pub event_type: EventType,
    pub direction: Direction,
    pub operation: String,       // e.g., "webhook", "get_diff", "create_batch"
    pub data: serde_json::Value, // Sanitized request/response data
    pub metadata: HashMap<String, String>, // PR number, repo, etc.
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
pub enum EventType {
    WebhookReceived,
    GitHubApiCall,
    OpenAiApiCall,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub enum Direction {
    Request,
    Response,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub enum ServiceType {
    GitHub,
    OpenAi,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct HttpEvent {
    pub method: String,
    pub url: String,
    pub headers: HashMap<String, String>, // Sanitized headers
    pub body: Option<serde_json::Value>,
    pub status_code: Option<u16>, // Only for responses
}

// Correlation ID type for better type safety
#[derive(Clone, Debug)]
pub struct CorrelationId(pub String);

// Header name for correlation ID propagation
pub const CORRELATION_ID_HEADER: &str = "X-Correlation-ID";
