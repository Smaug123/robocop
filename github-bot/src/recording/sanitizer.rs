use serde_json::Value;
use std::collections::HashMap;

pub struct Sanitizer;

impl Sanitizer {
    /// Remove sensitive data from headers
    pub fn sanitize_headers(headers: &HashMap<String, String>) -> HashMap<String, String> {
        let mut sanitized = HashMap::new();

        for (key, value) in headers {
            let key_lower = key.to_lowercase();
            let sanitized_value = match key_lower.as_str() {
                "authorization" => "[REDACTED]".to_string(),
                "x-hub-signature-256" => "[REDACTED]".to_string(),
                "cookie" => "[REDACTED]".to_string(),
                _ => value.clone(),
            };
            sanitized.insert(key.clone(), sanitized_value);
        }

        sanitized
    }

    /// Remove sensitive data from JSON payloads
    pub fn sanitize_json(value: &Value) -> Value {
        match value {
            Value::Object(map) => {
                let mut sanitized = serde_json::Map::new();
                for (key, val) in map {
                    let sanitized_val = match key.as_str() {
                        "token" | "private_key" | "secret" | "password" => {
                            Value::String("[REDACTED]".to_string())
                        }
                        _ => Self::sanitize_json(val),
                    };
                    sanitized.insert(key.clone(), sanitized_val);
                }
                Value::Object(sanitized)
            }
            Value::Array(arr) => Value::Array(arr.iter().map(Self::sanitize_json).collect()),
            _ => value.clone(),
        }
    }
}
