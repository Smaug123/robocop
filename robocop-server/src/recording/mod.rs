pub mod logger;
pub mod middleware;
pub mod sanitizer;
pub mod test_utils;
pub mod types;

pub use logger::RecordingLogger;
pub use middleware::RecordingMiddleware;
pub use types::*;
