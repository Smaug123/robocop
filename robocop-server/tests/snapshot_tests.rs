use insta::{assert_debug_snapshot, with_settings};
use robocop_server::recording::test_utils::{MockGenerator, TestSuite};
use robocop_server::recording::types::{Direction, EventType, RecordedEvent};
use serde_json::json;
use std::fs;

const RECORDINGS_PATHS: &[&str] = &["tests/recordings/sample1.jsonl"];

fn get_available_recording_paths() -> Vec<String> {
    let mut paths = Vec::new();

    for &base_path in RECORDINGS_PATHS {
        if std::path::Path::new(base_path).exists() {
            paths.push(base_path.to_string());
        }
    }

    // Also look for any directories that start with "recordings"
    if let Ok(entries) = fs::read_dir(".") {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.is_dir() {
                if let Some(name) = path.file_name().and_then(|n| n.to_str()) {
                    if name.starts_with("recordings") && !paths.contains(&name.to_string()) {
                        paths.push(name.to_string());
                    }
                }
            }
        }
    }

    paths
}

fn test_webhook_events_for_path(recordings_path: &str) {
    let test_suite = MockGenerator::generate_mocks_from_log(recordings_path)
        .expect("Failed to generate test suite from recordings");

    // Create snapshot for webhook events only
    let webhook_events: Vec<&RecordedEvent> = test_suite
        .scenarios
        .iter()
        .flat_map(|scenario| &scenario.events)
        .filter(|event| matches!(event.event_type, EventType::WebhookReceived))
        .collect();

    let snapshot_data = create_webhook_snapshot(&webhook_events, recordings_path);

    // Sanitize the path for snapshot name
    let snapshot_name = format!(
        "webhook_events_{}",
        recordings_path.replace(['/', '-'], "_")
    );

    with_settings!({
        description => format!("Webhook events from recordings: {}", recordings_path),
        omit_expression => true
    }, {
        assert_debug_snapshot!(snapshot_name, snapshot_data);
    });
}

#[tokio::test]
async fn test_webhook_events_snapshot() {
    let recording_paths = get_available_recording_paths();
    for path in recording_paths {
        test_webhook_events_for_path(&path);
    }
}

fn test_github_api_calls_for_path(recordings_path: &str) {
    let test_suite = MockGenerator::generate_mocks_from_log(recordings_path)
        .expect("Failed to generate test suite from recordings");

    let github_events: Vec<&RecordedEvent> = test_suite
        .scenarios
        .iter()
        .flat_map(|scenario| &scenario.events)
        .filter(|event| matches!(event.event_type, EventType::GitHubApiCall))
        .collect();

    let snapshot_data = create_api_calls_snapshot(&github_events, "GitHub", recordings_path);

    let snapshot_name = format!(
        "github_api_calls_{}",
        recordings_path.replace(['/', '-'], "_")
    );

    with_settings!({
        description => format!("GitHub API calls from recordings: {}", recordings_path),
        omit_expression => true
    }, {
        assert_debug_snapshot!(snapshot_name, snapshot_data);
    });
}

#[tokio::test]
async fn test_github_api_calls_snapshot() {
    let recording_paths = get_available_recording_paths();
    for path in recording_paths {
        test_github_api_calls_for_path(&path);
    }
}

fn test_openai_api_calls_for_path(recordings_path: &str) {
    let test_suite = MockGenerator::generate_mocks_from_log(recordings_path)
        .expect("Failed to generate test suite from recordings");

    let openai_events: Vec<&RecordedEvent> = test_suite
        .scenarios
        .iter()
        .flat_map(|scenario| &scenario.events)
        .filter(|event| matches!(event.event_type, EventType::OpenAiApiCall))
        .collect();

    let snapshot_data = create_api_calls_snapshot(&openai_events, "OpenAI", recordings_path);

    let snapshot_name = format!(
        "openai_api_calls_{}",
        recordings_path.replace(['/', '-'], "_")
    );

    with_settings!({
        description => format!("OpenAI API calls from recordings: {}", recordings_path),
        omit_expression => true
    }, {
        assert_debug_snapshot!(snapshot_name, snapshot_data);
    });
}

#[tokio::test]
async fn test_openai_api_calls_snapshot() {
    let recording_paths = get_available_recording_paths();
    for path in recording_paths {
        test_openai_api_calls_for_path(&path);
    }
}

fn test_complete_scenarios_for_path(recordings_path: &str) {
    let test_suite = MockGenerator::generate_mocks_from_log(recordings_path)
        .expect("Failed to generate test suite from recordings");

    // Create a snapshot of complete scenarios grouped by correlation_id
    let snapshot_data = create_scenarios_snapshot(&test_suite, recordings_path);

    let snapshot_name = format!(
        "complete_scenarios_{}",
        recordings_path.replace(['/', '-'], "_")
    );

    with_settings!({
        description => format!("Complete scenarios from recordings: {}", recordings_path),
        omit_expression => true
    }, {
        assert_debug_snapshot!(snapshot_name, snapshot_data);
    });
}

#[tokio::test]
async fn test_complete_scenarios_snapshot() {
    let recording_paths = get_available_recording_paths();
    for path in recording_paths {
        test_complete_scenarios_for_path(&path);
    }
}

fn test_request_response_pairs_for_path(recordings_path: &str) {
    let test_suite = MockGenerator::generate_mocks_from_log(recordings_path)
        .expect("Failed to generate test suite from recordings");

    // Extract request/response pairs for validation
    let pairs = extract_request_response_pairs(&test_suite);

    let snapshot_name = format!(
        "request_response_pairs_{}",
        recordings_path.replace(['/', '-'], "_")
    );

    with_settings!({
        description => format!("Request/response pairs from recordings: {}", recordings_path),
        omit_expression => true
    }, {
        assert_debug_snapshot!(snapshot_name, json!(pairs));
    });
}

#[tokio::test]
async fn test_request_response_pairs_snapshot() {
    let recording_paths = get_available_recording_paths();
    for path in recording_paths {
        test_request_response_pairs_for_path(&path);
    }
}

fn create_webhook_snapshot(events: &[&RecordedEvent], recordings_path: &str) -> serde_json::Value {
    let mut webhook_data: Vec<serde_json::Value> = events
        .iter()
        .map(|event| {
            json!({
                "correlation_id": event.correlation_id,
                "timestamp": event.timestamp,
                "operation": event.operation,
                "event_summary": extract_webhook_summary(&event.data),
                "sanitized_data": sanitize_webhook_data(&event.data)
            })
        })
        .collect();

    // Sort by timestamp for deterministic output
    webhook_data.sort_by(|a, b| {
        a["timestamp"]
            .as_str()
            .unwrap()
            .cmp(b["timestamp"].as_str().unwrap())
    });

    json!({
        "source": recordings_path,
        "total_webhook_events": webhook_data.len(),
        "events": webhook_data
    })
}

fn create_api_calls_snapshot(
    events: &[&RecordedEvent],
    service_name: &str,
    recordings_path: &str,
) -> serde_json::Value {
    let mut requests = Vec::new();
    let mut responses = Vec::new();

    for event in events {
        match event.direction {
            Direction::Request => {
                requests.push(json!({
                    "correlation_id": event.correlation_id,
                    "timestamp": event.timestamp,
                    "operation": event.operation,
                    "request_summary": extract_api_request_summary(&event.data),
                    "sanitized_data": sanitize_api_data(&event.data)
                }));
            }
            Direction::Response => {
                responses.push(json!({
                    "correlation_id": event.correlation_id,
                    "timestamp": event.timestamp,
                    "operation": event.operation,
                    "response_summary": extract_api_response_summary(&event.data),
                    "sanitized_data": sanitize_api_data(&event.data)
                }));
            }
        }
    }

    // Sort by timestamp for deterministic output
    requests.sort_by(|a, b| {
        a["timestamp"]
            .as_str()
            .unwrap()
            .cmp(b["timestamp"].as_str().unwrap())
    });
    responses.sort_by(|a, b| {
        a["timestamp"]
            .as_str()
            .unwrap()
            .cmp(b["timestamp"].as_str().unwrap())
    });

    json!({
        "source": recordings_path,
        "service": service_name,
        "total_requests": requests.len(),
        "total_responses": responses.len(),
        "requests": requests,
        "responses": responses
    })
}

fn create_scenarios_snapshot(test_suite: &TestSuite, recordings_path: &str) -> serde_json::Value {
    let mut scenarios: Vec<serde_json::Value> = test_suite
        .scenarios
        .iter()
        .map(|scenario| {
            let mut events: Vec<serde_json::Value> = scenario
                .events
                .iter()
                .map(|event| {
                    json!({
                        "timestamp": event.timestamp,
                        "event_type": format!("{:?}", event.event_type),
                        "direction": format!("{:?}", event.direction),
                        "operation": event.operation,
                        "summary": create_event_summary(event),
                        "sanitized_data": sanitize_event_data(event)
                    })
                })
                .collect();

            // Sort events by timestamp
            events.sort_by(|a, b| {
                a["timestamp"].as_str().unwrap().cmp(b["timestamp"].as_str().unwrap())
            });

            json!({
                "scenario_name": scenario.name,
                "correlation_id": scenario.events.first().map(|e| &e.correlation_id),
                "event_count": scenario.events.len(),
                "event_types": scenario.events.iter().map(|e| format!("{:?}", e.event_type)).collect::<Vec<_>>(),
                "events": events
            })
        })
        .collect();

    // Sort scenarios by correlation_id for deterministic output
    scenarios.sort_by(|a, b| {
        a["correlation_id"]
            .as_str()
            .unwrap()
            .cmp(b["correlation_id"].as_str().unwrap())
    });

    json!({
        "source": recordings_path,
        "total_scenarios": scenarios.len(),
        "scenarios": scenarios
    })
}

fn extract_request_response_pairs(test_suite: &TestSuite) -> Vec<serde_json::Value> {
    let mut pairs = Vec::new();

    for scenario in &test_suite.scenarios {
        let requests: Vec<&RecordedEvent> = scenario
            .events
            .iter()
            .filter(|e| matches!(e.direction, Direction::Request))
            .collect();

        for request in requests {
            if let Some(response) = scenario.events.iter().find(|e| {
                matches!(e.direction, Direction::Response)
                    && e.event_type == request.event_type
                    && e.correlation_id == request.correlation_id
            }) {
                pairs.push(json!({
                    "correlation_id": request.correlation_id,
                    "operation": request.operation,
                    "event_type": format!("{:?}", request.event_type),
                    "request": {
                        "timestamp": request.timestamp,
                        "summary": create_event_summary(request),
                        "sanitized_data": sanitize_event_data(request)
                    },
                    "response": {
                        "timestamp": response.timestamp,
                        "summary": create_event_summary(response),
                        "sanitized_data": sanitize_event_data(response)
                    }
                }));
            }
        }
    }

    // Sort pairs by request timestamp for deterministic output
    pairs.sort_by(|a, b| {
        a["request"]["timestamp"]
            .as_str()
            .unwrap()
            .cmp(b["request"]["timestamp"].as_str().unwrap())
    });

    pairs
}

fn extract_webhook_summary(data: &serde_json::Value) -> serde_json::Value {
    if let Some(body) = data.get("body") {
        json!({
            "action": body.get("action"),
            "ref": body.get("ref"),
            "repository_name": body.get("repository").and_then(|r| r.get("name")),
            "sender": body.get("sender").and_then(|s| s.get("login")),
            "commits_count": body.get("commits").and_then(|c| c.as_array()).map(|a| a.len())
        })
    } else {
        json!({})
    }
}

fn extract_api_request_summary(data: &serde_json::Value) -> serde_json::Value {
    json!({
        "method": data.get("method"),
        "url_pattern": extract_url_pattern(data.get("url")),
        "has_body": data.get("body").is_some()
    })
}

fn extract_api_response_summary(data: &serde_json::Value) -> serde_json::Value {
    json!({
        "status_code": data.get("status_code"),
        "has_body": data.get("body").is_some(),
        "content_type": data.get("headers").and_then(|h| h.get("content-type"))
    })
}

fn extract_url_pattern(url: Option<&serde_json::Value>) -> Option<String> {
    url.and_then(|u| u.as_str()).map(|url_str| {
        if let Ok(parsed) = url::Url::parse(url_str) {
            format!("{}{}", parsed.host_str().unwrap_or(""), parsed.path())
        } else {
            url_str.to_string()
        }
    })
}

fn create_event_summary(event: &RecordedEvent) -> serde_json::Value {
    match event.event_type {
        EventType::WebhookReceived => extract_webhook_summary(&event.data),
        EventType::GitHubApiCall | EventType::OpenAiApiCall => match event.direction {
            Direction::Request => extract_api_request_summary(&event.data),
            Direction::Response => extract_api_response_summary(&event.data),
        },
    }
}

fn sanitize_webhook_data(data: &serde_json::Value) -> serde_json::Value {
    let mut sanitized = data.clone();

    // Remove/redact sensitive information in headers
    if let Some(headers) = sanitized.get_mut("headers") {
        if let Some(headers_obj) = headers.as_object_mut() {
            // Redact sensitive headers
            for sensitive_header in ["x-hub-signature-256", "authorization", "x-github-token"] {
                if headers_obj.contains_key(sensitive_header) {
                    headers_obj.insert(sensitive_header.to_string(), json!("[REDACTED]"));
                }
            }
        }
    }

    // Keep structure but remove large nested objects for clarity
    if let Some(body) = sanitized.get_mut("body") {
        if let Some(repo) = body.get_mut("repository") {
            if let Some(repo_obj) = repo.as_object_mut() {
                // Keep only essential repo fields
                let essential_fields = ["id", "name", "full_name", "private", "default_branch"];
                repo_obj.retain(|k, _| essential_fields.contains(&k.as_str()));
            }
        }
    }

    sanitized
}

fn sanitize_api_data(data: &serde_json::Value) -> serde_json::Value {
    let mut sanitized = data.clone();

    // Redact authorization headers
    if let Some(headers) = sanitized.get_mut("headers") {
        if let Some(headers_obj) = headers.as_object_mut() {
            for auth_header in ["authorization", "x-github-token", "openai-api-key"] {
                if headers_obj.contains_key(auth_header) {
                    headers_obj.insert(auth_header.to_string(), json!("[REDACTED]"));
                }
            }
        }
    }

    // For large response bodies, summarize instead of including full content
    if let Some(body) = sanitized.get_mut("body") {
        if let Some(body_str) = body.as_str() {
            if body_str.len() > 1000 {
                let length = body_str.len();
                sanitized.as_object_mut().unwrap().insert(
                    "body".to_string(),
                    json!(format!("[LARGE_BODY: {} chars]", length)),
                );
            }
        }
    }

    sanitized
}

fn sanitize_event_data(event: &RecordedEvent) -> serde_json::Value {
    match event.event_type {
        EventType::WebhookReceived => sanitize_webhook_data(&event.data),
        EventType::GitHubApiCall | EventType::OpenAiApiCall => sanitize_api_data(&event.data),
    }
}
