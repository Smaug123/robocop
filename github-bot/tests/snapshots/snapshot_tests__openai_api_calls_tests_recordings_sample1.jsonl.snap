---
source: tests/snapshot_tests.rs
description: "OpenAI API calls from recordings: tests/recordings/sample1.jsonl"
---
Object {
    "requests": Array [
        Object {
            "correlation_id": String("14204c3a-581a-4c17-951c-cbfcd1718143"),
            "operation": String("POST /v1/batches"),
            "request_summary": Object {
                "has_body": Bool(true),
                "method": String("POST"),
                "url_pattern": String("api.openai.com/v1/batches"),
            },
            "sanitized_data": Object {
                "body": String("{\"input_file_id\":\"file-QHK9Xmx8zLueuQukq5B62L\",\"endpoint\":\"/v1/chat/completions\",\"completion_window\":\"24h\",\"metadata\":{\"repo_name\":\"robocop\",\"metadata_schema\":\"1\",\"target_commit\":\"5a1d0c69943dd005a052013db8fd0bc133a6eafa\",\"pull_request_url\":\"https://github.com/Smaug123/robocop/pull/8\",\"version\":\"c031981f\",\"source_commit\":\"e6a8b9636297ffc84c63193d18a38dea197867f3\",\"description\":\"robocop code review tool\",\"branch\":\"cope-with-pr-url\"}}"),
                "headers": Object {
                    "authorization": String("[REDACTED]"),
                    "content-type": String("application/json"),
                },
                "method": String("POST"),
                "url": String("https://api.openai.com/v1/batches"),
            },
            "timestamp": String("2025-10-12T07:37:59.679595+00:00"),
        },
        Object {
            "correlation_id": String("cd298d30-fe77-49cb-a506-f67f9cf549c5"),
            "operation": String("GET /v1/batches/batch_68eb5ad7f45881908195fbb16a5d7fdf"),
            "request_summary": Object {
                "has_body": Bool(true),
                "method": String("GET"),
                "url_pattern": String("api.openai.com/v1/batches/batch_68eb5ad7f45881908195fbb16a5d7fdf"),
            },
            "sanitized_data": Object {
                "body": String("[NO_BODY]"),
                "headers": Object {
                    "authorization": String("[REDACTED]"),
                    "content-type": String("application/json"),
                },
                "method": String("GET"),
                "url": String("https://api.openai.com/v1/batches/batch_68eb5ad7f45881908195fbb16a5d7fdf"),
            },
            "timestamp": String("2025-10-12T07:38:37.473901+00:00"),
        },
        Object {
            "correlation_id": String("e9f35baa-37fc-4d14-a8c5-67af5a27254d"),
            "operation": String("GET /v1/batches/batch_68eb5ad7f45881908195fbb16a5d7fdf"),
            "request_summary": Object {
                "has_body": Bool(true),
                "method": String("GET"),
                "url_pattern": String("api.openai.com/v1/batches/batch_68eb5ad7f45881908195fbb16a5d7fdf"),
            },
            "sanitized_data": Object {
                "body": String("[NO_BODY]"),
                "headers": Object {
                    "authorization": String("[REDACTED]"),
                    "content-type": String("application/json"),
                },
                "method": String("GET"),
                "url": String("https://api.openai.com/v1/batches/batch_68eb5ad7f45881908195fbb16a5d7fdf"),
            },
            "timestamp": String("2025-10-12T07:39:37.472542+00:00"),
        },
        Object {
            "correlation_id": String("32127334-af13-4bc0-9209-120a86c038c7"),
            "operation": String("GET /v1/batches/batch_68eb5ad7f45881908195fbb16a5d7fdf"),
            "request_summary": Object {
                "has_body": Bool(true),
                "method": String("GET"),
                "url_pattern": String("api.openai.com/v1/batches/batch_68eb5ad7f45881908195fbb16a5d7fdf"),
            },
            "sanitized_data": Object {
                "body": String("[NO_BODY]"),
                "headers": Object {
                    "authorization": String("[REDACTED]"),
                    "content-type": String("application/json"),
                },
                "method": String("GET"),
                "url": String("https://api.openai.com/v1/batches/batch_68eb5ad7f45881908195fbb16a5d7fdf"),
            },
            "timestamp": String("2025-10-12T07:40:37.467660+00:00"),
        },
        Object {
            "correlation_id": String("b4b2064d-329f-4dc7-a75c-b770d5492396"),
            "operation": String("GET /v1/batches/batch_68eb5ad7f45881908195fbb16a5d7fdf"),
            "request_summary": Object {
                "has_body": Bool(true),
                "method": String("GET"),
                "url_pattern": String("api.openai.com/v1/batches/batch_68eb5ad7f45881908195fbb16a5d7fdf"),
            },
            "sanitized_data": Object {
                "body": String("[NO_BODY]"),
                "headers": Object {
                    "authorization": String("[REDACTED]"),
                    "content-type": String("application/json"),
                },
                "method": String("GET"),
                "url": String("https://api.openai.com/v1/batches/batch_68eb5ad7f45881908195fbb16a5d7fdf"),
            },
            "timestamp": String("2025-10-12T07:41:37.467027+00:00"),
        },
        Object {
            "correlation_id": String("2fa4f735-15d1-43d7-b630-57dbd272fb90"),
            "operation": String("GET /v1/files/file-4HBYcxyKe3vaLvisdW2hcG/content"),
            "request_summary": Object {
                "has_body": Bool(true),
                "method": String("GET"),
                "url_pattern": String("api.openai.com/v1/files/file-4HBYcxyKe3vaLvisdW2hcG/content"),
            },
            "sanitized_data": Object {
                "body": String("[NO_BODY]"),
                "headers": Object {
                    "authorization": String("[REDACTED]"),
                },
                "method": String("GET"),
                "url": String("https://api.openai.com/v1/files/file-4HBYcxyKe3vaLvisdW2hcG/content"),
            },
            "timestamp": String("2025-10-12T07:41:37.710093+00:00"),
        },
    ],
    "responses": Array [
        Object {
            "correlation_id": String("14204c3a-581a-4c17-951c-cbfcd1718143"),
            "operation": String("response_200"),
            "response_summary": Object {
                "content_type": String("application/json"),
                "has_body": Bool(false),
                "status_code": Number(200),
            },
            "sanitized_data": Object {
                "body_size": Number(1204),
                "headers": Object {
                    "alt-svc": String("h3=\":443\"; ma=86400"),
                    "cf-cache-status": String("DYNAMIC"),
                    "cf-ray": String("98d4ef647cf31484-LHR"),
                    "connection": String("keep-alive"),
                    "content-length": String("1204"),
                    "content-type": String("application/json"),
                    "date": String("Sun, 12 Oct 2025 07:38:00 GMT"),
                    "openai-organization": String("user-unqgrcghgdmbulnwwgwgiyqb"),
                    "openai-processing-ms": String("380"),
                    "openai-project": String("proj_6wTwiRUWJMtX13kVZHqQWEK0"),
                    "openai-version": String("2020-10-01"),
                    "server": String("cloudflare"),
                    "set-cookie": String("_cfuvid=VxKzUdrleFs5O6mCsSsIFdiWTVQDF4Fw7EQ.5ZGcqR4-1760254680314-0.0.1.1-604800000; path=/; domain=.api.openai.com; HttpOnly; Secure; SameSite=None"),
                    "strict-transport-security": String("max-age=31536000; includeSubDomains; preload"),
                    "x-content-type-options": String("nosniff"),
                    "x-envoy-upstream-service-time": String("382"),
                    "x-openai-proxy-wasm": String("v0.1"),
                    "x-request-id": String("req_134804a9178948548b0ab83e4fb3b494"),
                },
                "status_code": Number(200),
            },
            "timestamp": String("2025-10-12T07:38:00.369193+00:00"),
        },
        Object {
            "correlation_id": String("cd298d30-fe77-49cb-a506-f67f9cf549c5"),
            "operation": String("response_200"),
            "response_summary": Object {
                "content_type": String("application/json"),
                "has_body": Bool(false),
                "status_code": Number(200),
            },
            "sanitized_data": Object {
                "body_size": Number(1204),
                "headers": Object {
                    "alt-svc": String("h3=\":443\"; ma=86400"),
                    "cf-cache-status": String("DYNAMIC"),
                    "cf-ray": String("98d4f05078411484-LHR"),
                    "connection": String("keep-alive"),
                    "content-length": String("1204"),
                    "content-type": String("application/json"),
                    "date": String("Sun, 12 Oct 2025 07:38:37 GMT"),
                    "openai-organization": String("user-unqgrcghgdmbulnwwgwgiyqb"),
                    "openai-processing-ms": String("27"),
                    "openai-project": String("proj_6wTwiRUWJMtX13kVZHqQWEK0"),
                    "openai-version": String("2020-10-01"),
                    "server": String("cloudflare"),
                    "set-cookie": String("_cfuvid=aVxpxg.HAVczlwtEY13DoCALoEpz2vIEr8EWBFSMKyY-1760254717696-0.0.1.1-604800000; path=/; domain=.api.openai.com; HttpOnly; Secure; SameSite=None"),
                    "strict-transport-security": String("max-age=31536000; includeSubDomains; preload"),
                    "x-content-type-options": String("nosniff"),
                    "x-envoy-upstream-service-time": String("29"),
                    "x-openai-proxy-wasm": String("v0.1"),
                    "x-request-id": String("req_5f4d30dfe9de49b3bb0bf83639a5b4de"),
                },
                "status_code": Number(200),
            },
            "timestamp": String("2025-10-12T07:38:37.662220+00:00"),
        },
        Object {
            "correlation_id": String("e9f35baa-37fc-4d14-a8c5-67af5a27254d"),
            "operation": String("response_200"),
            "response_summary": Object {
                "content_type": String("application/json"),
                "has_body": Bool(false),
                "status_code": Number(200),
            },
            "sanitized_data": Object {
                "body_size": Number(1225),
                "headers": Object {
                    "alt-svc": String("h3=\":443\"; ma=86400"),
                    "cf-cache-status": String("DYNAMIC"),
                    "cf-ray": String("98d4f1c789941484-LHR"),
                    "connection": String("keep-alive"),
                    "content-length": String("1225"),
                    "content-type": String("application/json"),
                    "date": String("Sun, 12 Oct 2025 07:39:37 GMT"),
                    "openai-organization": String("user-unqgrcghgdmbulnwwgwgiyqb"),
                    "openai-processing-ms": String("28"),
                    "openai-project": String("proj_6wTwiRUWJMtX13kVZHqQWEK0"),
                    "openai-version": String("2020-10-01"),
                    "server": String("cloudflare"),
                    "set-cookie": String("_cfuvid=suaidMw6.pCkblIft0NSJ_AcGLHGBppKhT02AQZ0R7U-1760254777701-0.0.1.1-604800000; path=/; domain=.api.openai.com; HttpOnly; Secure; SameSite=None"),
                    "strict-transport-security": String("max-age=31536000; includeSubDomains; preload"),
                    "x-content-type-options": String("nosniff"),
                    "x-envoy-upstream-service-time": String("30"),
                    "x-openai-proxy-wasm": String("v0.1"),
                    "x-request-id": String("req_da6214da0b374b3c8d34db8c33803a81"),
                },
                "status_code": Number(200),
            },
            "timestamp": String("2025-10-12T07:39:37.666911+00:00"),
        },
        Object {
            "correlation_id": String("32127334-af13-4bc0-9209-120a86c038c7"),
            "operation": String("response_200"),
            "response_summary": Object {
                "content_type": String("application/json"),
                "has_body": Bool(false),
                "status_code": Number(200),
            },
            "sanitized_data": Object {
                "body_size": Number(1225),
                "headers": Object {
                    "alt-svc": String("h3=\":443\"; ma=86400"),
                    "cf-cache-status": String("DYNAMIC"),
                    "cf-ray": String("98d4f33e8d601484-LHR"),
                    "connection": String("keep-alive"),
                    "content-length": String("1225"),
                    "content-type": String("application/json"),
                    "date": String("Sun, 12 Oct 2025 07:40:37 GMT"),
                    "openai-organization": String("user-unqgrcghgdmbulnwwgwgiyqb"),
                    "openai-processing-ms": String("25"),
                    "openai-project": String("proj_6wTwiRUWJMtX13kVZHqQWEK0"),
                    "openai-version": String("2020-10-01"),
                    "server": String("cloudflare"),
                    "set-cookie": String("_cfuvid=w9XHnLpDCZRwm.LGc9bNJwqiHdvHsvg8DtSu.A_zI3U-1760254837706-0.0.1.1-604800000; path=/; domain=.api.openai.com; HttpOnly; Secure; SameSite=None"),
                    "strict-transport-security": String("max-age=31536000; includeSubDomains; preload"),
                    "x-content-type-options": String("nosniff"),
                    "x-envoy-upstream-service-time": String("27"),
                    "x-openai-proxy-wasm": String("v0.1"),
                    "x-request-id": String("req_b236651ef0294835b880f1f58dcf881a"),
                },
                "status_code": Number(200),
            },
            "timestamp": String("2025-10-12T07:40:37.666985+00:00"),
        },
        Object {
            "correlation_id": String("b4b2064d-329f-4dc7-a75c-b770d5492396"),
            "operation": String("response_200"),
            "response_summary": Object {
                "content_type": String("application/json"),
                "has_body": Bool(false),
                "status_code": Number(200),
            },
            "sanitized_data": Object {
                "body_size": Number(1274),
                "headers": Object {
                    "alt-svc": String("h3=\":443\"; ma=86400"),
                    "cf-cache-status": String("DYNAMIC"),
                    "cf-ray": String("98d4f4b59f951484-LHR"),
                    "connection": String("keep-alive"),
                    "content-length": String("1274"),
                    "content-type": String("application/json"),
                    "date": String("Sun, 12 Oct 2025 07:41:37 GMT"),
                    "openai-organization": String("user-unqgrcghgdmbulnwwgwgiyqb"),
                    "openai-processing-ms": String("43"),
                    "openai-project": String("proj_6wTwiRUWJMtX13kVZHqQWEK0"),
                    "openai-version": String("2020-10-01"),
                    "server": String("cloudflare"),
                    "set-cookie": String("_cfuvid=X3JdXRRPZ_.v35fS9tC8HO0MyllKDhawSJfJ8wEGZAY-1760254897750-0.0.1.1-604800000; path=/; domain=.api.openai.com; HttpOnly; Secure; SameSite=None"),
                    "strict-transport-security": String("max-age=31536000; includeSubDomains; preload"),
                    "x-content-type-options": String("nosniff"),
                    "x-envoy-upstream-service-time": String("70"),
                    "x-openai-proxy-wasm": String("v0.1"),
                    "x-request-id": String("req_eea5ea284d384000bb138c78a9d5c9ef"),
                },
                "status_code": Number(200),
            },
            "timestamp": String("2025-10-12T07:41:37.709843+00:00"),
        },
        Object {
            "correlation_id": String("2fa4f735-15d1-43d7-b630-57dbd272fb90"),
            "operation": String("response_200"),
            "response_summary": Object {
                "content_type": String("application/octet-stream"),
                "has_body": Bool(false),
                "status_code": Number(200),
            },
            "sanitized_data": Object {
                "body_size": Number(3189),
                "headers": Object {
                    "access-control-allow-origin": String("*"),
                    "alt-svc": String("h3=\":443\"; ma=86400"),
                    "cf-cache-status": String("DYNAMIC"),
                    "cf-ray": String("98d4f4b70a4f1484-LHR"),
                    "connection": String("keep-alive"),
                    "content-disposition": String("attachment; filename=\"batch_68eb5ad7f45881908195fbb16a5d7fdf_output.jsonl\""),
                    "content-length": String("3189"),
                    "content-type": String("application/octet-stream"),
                    "date": String("Sun, 12 Oct 2025 07:41:38 GMT"),
                    "openai-organization": String("user-unqgrcghgdmbulnwwgwgiyqb"),
                    "openai-processing-ms": String("95"),
                    "openai-project": String("proj_6wTwiRUWJMtX13kVZHqQWEK0"),
                    "openai-version": String("2020-10-01"),
                    "server": String("cloudflare"),
                    "set-cookie": String("_cfuvid=Wer8tavSD_e8dFKPb6N28A7Qg8kONrRPWz7VcWTlPdY-1760254898016-0.0.1.1-604800000; path=/; domain=.api.openai.com; HttpOnly; Secure; SameSite=None"),
                    "strict-transport-security": String("max-age=31536000; includeSubDomains; preload"),
                    "x-content-type-options": String("nosniff"),
                    "x-envoy-upstream-service-time": String("103"),
                    "x-openai-proxy-wasm": String("v0.1"),
                    "x-request-id": String("req_03e1779bcd3b7fdab61181c1d31957f7"),
                },
                "status_code": Number(200),
            },
            "timestamp": String("2025-10-12T07:41:37.975837+00:00"),
        },
    ],
    "service": String("OpenAI"),
    "source": String("tests/recordings/sample1.jsonl"),
    "total_requests": Number(6),
    "total_responses": Number(6),
}
